src/main/scala/org/constellation/consensus/Round.scala

```diff
@@ -0,0 +1,157 @@
+import org.constellation.primitives._
+import org.constellation.util.Validation.EnrichedFuture
+import org.constellation.{ConfigUtil, DAO}
+
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.ExecutionContextExecutor
+import scala.concurrent.duration._
+
+class Round(id: RoundId, dao: DAO) extends Actor with ActorLogging {
+
+  val transactionProposals: mutable.Map[FacilitatorId, Seq[Transaction]] = mutable.Map()
+  val checkpointBlockProposals: mutable.Map[FacilitatorId, CheckpointBlock] = mutable.Map()
+
+  var numberOfCheckpointBlockProposalsSent: Int = _
+  var numberOfTransactionProposalsSent: Int = _
+
+  var unionTransactionProposalsTikTok: Cancellable = _
+
+
+  override def preStart(): Unit = {
+    self ! Initialize
+  }
+
+  override def receive: Receive = {
+
+    case Initialize =>
+      dao
+        .pullTransactions(dao.minCheckpointFormationThreshold)
+        .foreach { transactions =>
+          transactionProposals += FacilitatorId(dao.id) -> transactions
+          val readyPeers = dao.readyPeers
+          context.parent ! TransactionsProposal(id,
+                                                readyPeers.values,
+                                                FacilitatorId(dao.id),
+                                                transactions)
+          numberOfTransactionProposalsSent = readyPeers.size + 1
+          unionTransactionProposalsTikTok = scheduleUnionProposals
+        }
+    case ReceivedProposal(_, facilitatorId, transactions) =>
+      transactionProposals += facilitatorId -> transactions
+      if (transactionProposals.size >= numberOfTransactionProposalsSent) {
+        Option(unionTransactionProposalsTikTok).foreach(_.cancel())
+        self ! UnionProposals(id)
+      }
+    case ReceivedUnionedBlock(_, facilitatorId, checkpointBlock) =>
+      checkpointBlockProposals += facilitatorId -> checkpointBlock
+      if (checkpointBlockProposals.size >= numberOfCheckpointBlockProposalsSent) {
+        self ! ResolveMajorityCheckpointBlock(id)
+      }
+    case UnionProposals(_)                 => unionProposals()
+    case ResolveMajorityCheckpointBlock(_) => resolveMajorityCheckpointBlock()
+    case msg                               => log.info(s"Received unknown message: $msg")
+  }
+
+  private def scheduleUnionProposals: Cancellable =
+    context.system.scheduler.scheduleOnce(
+    ConfigUtil.getDurationFromConfig("constellation.consensus.union-proposals-timeout", 15.second),
+    self,
+    UnionProposals(id)
+  )
+
+
+  def unionProposals(): Unit = {
+    // TODO: wkoszycki is this the way to merge transactions ?
+    val transactions = transactionProposals.values.flatten.toSet
+    // TODO: wkoszycki should I pull messages and tips this way (services are altering the data itself)?
+    val messages = dao.threadSafeMessageMemPool.pull(1).getOrElse(Seq())
+    val maybeCheckpointBlock = dao
+      .pullTips()
```

Comment:
This can't happen at this stage. What will end up happening here is each node will create a different block with different tip references. What should happen instead is the node that triggers StartNewBlockCreationRound should pull tips there, and send that information (the facilitators chosen, and the tips selected,) to all the other facilitators, who should then submit proposals, this should happen on initialization of the Round class essentially